compiler-rt/lib/scudo/standalone/tests/secondary_test.cpp

```diff
@@ -265,3 +265,99 @@ TEST_F(MapAllocatorWithReleaseTest, SecondaryThreadsRace) {
   Allocator->getStats(&Str);
   Str.output();
 }
+
+struct MapAllocatorCacheTest : public Test {
+  static constexpr scudo::u32 MarkerBytes = 0xDEADBEEF;
+
+  static void testUnmapCallback(scudo::MemMapT &MemMap) {
+    scudo::u32 *Ptr = reinterpret_cast<scudo::u32 *>(MemMap.getBase());
+    *Ptr = MarkerBytes;
+  }
+
+  using Config = scudo::DefaultConfig;
+  using SecondaryConfig = scudo::SecondaryConfig<Config>;
+  using CacheConfig = SecondaryConfig::CacheConfig;
+  using CacheT = scudo::MapAllocatorCache<CacheConfig, testUnmapCallback>;
+
+  std::unique_ptr<CacheT> Cache = std::make_unique<CacheT>();
+
+  const scudo::uptr PageSize = scudo::getPageSizeCached();
+  static constexpr scudo::uptr DefaultAllocSize = scudo::uptr(1U) << 16;
+
+  scudo::Options Options = getOptionsForConfig<SecondaryConfig>();
+
+  void SetUp() override { Cache->init(/*ReleaseToOsInterval=*/-1); }
+
+  void TearDown() override { Cache->unmapTestOnly(); }
+
+  scudo::MemMapT allocate(scudo::uptr Size) {
+    scudo::uptr MapSize = scudo::roundUp(Size, PageSize);
+    scudo::ReservedMemoryT ReservedMemory;
+    CHECK(ReservedMemory.create(0U, MapSize, nullptr, MAP_ALLOWNOMEM));
+
+    scudo::MemMapT MemMap = ReservedMemory.dispatch(
+        ReservedMemory.getBase(), ReservedMemory.getCapacity());
+    MemMap.remap(MemMap.getBase(), MemMap.getCapacity(), "scudo:test",
+                 MAP_RESIZABLE | MAP_ALLOWNOMEM);
+    return MemMap;
+  }
+
+  void fillCacheWithSameSizeBlocks(std::vector<scudo::MemMapT> &MemMaps,
+                                   scudo::uptr NumEntries, scudo::uptr Size) {
+    for (scudo::uptr I = 0; I < NumEntries; I++)
+      MemMaps.emplace_back(allocate(Size));
+    for (scudo::uptr I = 0; I < NumEntries; I++) {
+      scudo::MemMapT &MemMap = MemMaps[I];
+      Cache->store(Options, MemMap.getBase(), MemMap.getCapacity(),
+                   MemMap.getBase(), MemMap);
+    }
+  }
+};
+
+TEST_F(MapAllocatorCacheTest, CacheOrder) {
+  std::vector<scudo::MemMapT> MemMaps;
```

Comment:
This might be better done with a class variable. Then you can move the clean up to the TearDown function and not have to worry about the copy pasting the clean-up code in all tests.

Reply:
I think @ChiaHungDuan  requested that the `MemMaps` variable be specific to the tests for now because if more tests are added in the future they may not use it.

Reply:
It was the class variable and my opinion is that this is used/needed for some tests but not all. About the clean up, I was wondering the same question then I think in most scudo tests, we usually ensure everything is returned/cleaned up before exit, for example, in primary test, we have to make sure all the blocks are released at the end so the verification in TearDown won't complain memory leak. It's also kind of a reminder for writing the scudo test that we need to know what we have created and what we have to do afterwards. Therefore, I suggested Jush to make the container local.

Reply:
That's fine with me. Since this is in only two tests, it's not too bad to copy it. But if you are adding more new tests with this, I would recommend the cleanup go in a cleanup function. Or, if you really wanted to be tricky, there is a way to do this with a vector of unique_ptrs and a custom destroy function. But that feels like overkill.

Reply:
An alternative may be a wrapper like ScopedMemMap and it will only be used in the test.

Reply:
> An alternative may be a wrapper like ScopedMemMap and it will only be used in the test.

That's possible but `MemMap` doesn't have a copy constructor nor an overloaded `=` operator right now, which I think we would need in making `ScopedMemMap` to support the vector of `ScopedMemMaps`.

Reply:
Discussed in meeting: for now we will leave it as is since all of the current tests follow the same logic for unmapping the MemMaps.